pallets/foreign-investments/src/impls/invest.rs

```diff
@@ -0,0 +1,1127 @@
+				..
+			} => Err(DispatchError::Other(
+				"Invalid invest state when transitioning an increased swap order with the same in- \
+				 and outgoing currency",
+			)),
+			_ => Err(DispatchError::Other(
+				"Invalid invest state, should automatically be transitioned into state without \
+				 AndSwapIntoForeignDone",
+			)),
+		}
+	}
+
+	/// Handle decrease transitions for the same incoming and outgoing
+	/// currencies.
+	///
+	/// NOTES:
+	/// * We can never directly compare `swap.amount` or `done_amount` with
+	///   `pool_swap.amount` and `invest_amount` if the currencies mismatch as
+	///   the former pair is denominated in foreign currency and the latter pair
+	///   in pool currency.
+	/// * We can ignore handling all states which include `SwapIntoForeignDone`
+	///   without `ActiveSwapIntoForeignCurrency` as we consume the done amount
+	///   and transition in the post transition phase. Moreover, we can ignore
+	///   any state which involves an active swap, i.e. `ActiveSwapInto{Pool,
+	///   Return}Currency`, as these must not exist if the in and out currency
+	///   is the same. To be safe and to not make any unhandled assumptions, we
+	///   throw `DispatchError::Other` for these states though we need to make
+	///   sure this can never occur!
+	fn handle_decrease_non_foreign(
+		&self,
+		swap: Swap<T::Balance, T::CurrencyId>,
+	) -> Result<Self, DispatchError> {
+		if let Self::InvestmentOngoing { invest_amount } = &self {
+			if swap.amount < *invest_amount {
+				Ok(Self::SwapIntoForeignDoneAndInvestmentOngoing {
+					done_swap: swap,
+					invest_amount: invest_amount.ensure_sub(swap.amount)?,
+				})
+			} else {
+				Ok(Self::SwapIntoForeignDone { done_swap: swap })
+			}
+		}
+		// should never occur but let's be safe here
+		else {
+			Err(DispatchError::Other(
+				"Invalid invest state when transitioning a decreased swap order with the same in- \
+				 and outgoing currency",
+			))
+		}
+	}
+
+	/// Update or kill the state's unprocessed investing amount.
+	/// * If the state includes `InvestmentOngoing`, either update or remove the
+	///   invested amount.
+	/// * Else the unprocessed amount should be zero. If it is not, state is
+	///   corrupted as this reflects the investment was increased improperly.
+	fn handle_collect(&self, unprocessed_amount: T::Balance) -> Result<Self, DispatchError> {
+		match self {
+			Self::InvestmentOngoing { .. } => {
+				if unprocessed_amount.is_zero() {
+					Ok(Self::NoState)
+				} else {
+					Ok(Self::InvestmentOngoing {
+						invest_amount: unprocessed_amount,
+					})
+				}
+			}
+			Self::ActiveSwapIntoPoolCurrencyAndInvestmentOngoing { swap, .. } => {
+				if unprocessed_amount.is_zero() {
+					Ok(Self::ActiveSwapIntoPoolCurrency { swap: *swap })
+				} else {
+					Ok(Self::ActiveSwapIntoPoolCurrencyAndInvestmentOngoing {
+						swap: *swap,
+						invest_amount: unprocessed_amount,
+					})
+				}
+			}
```

Comment:
Wondering if the previous state contains `InvestmentOngoing`, should then that value be added to the `unprocessed_amount` in line 1052?

```rust
invest_amount: previous_amount + unprocessed_amount
```

Reply:
I made the comment before you update the changes, but until I see it, I think it still applies.

Reply:
Since this is part of the `handle_collect` transition, the current implementation is correct. When an investment is collected, the unprocessed amount either decreases or stays the same if none of the investment was processed during epoch execution. 

We provide the currently unprocessed obtained by `T::Investment::investment(..)` as argument to `handle_collect` such that we can simply set the `invest_amount` to that value.

Reply:
> When an investment is collected, the unprocessed amount either decreases or stays

When exactly does it decrease at `handle_collect()`? I can not see any decrease. 

From my point of view, either we are overwriting the amount when not, and maybe it should be decreased, or we are storing `invest_amount` when we really do not need it because it can be obtained directly from `T::Investment::investment(..)`.

Not sure if I'm losing something 🥲

Reply:
Regarding your point 6. of https://centrifugedao.slack.com/archives/C04HCK98KLL/p1694526988803069?thread_ts=1694518146.332759&cid=C04HCK98KLL

> When an investment is collected, the invest_amount must be decreased exactly by the CollectedInvestment::<T>::get(&investor, investment_id).amount_payment

Should this be decreased then?

Reply:
> When exactly does it decrease at handle_collect()? I can not see any decrease.

The unprocessed investment amount decreases exactly by the collected amount. Say I have an unprocessed investment of 1000 USDC and 500 USDC gets collected. Then, `handle_collect(500)` is executed and thus the `invest_amount` set to 500, i.e. "decreased" from the previous 1000 to 500. Maybe _decrease_ is not the correct terminology here as we just reset the invest amount.

Reply:
> Regarding your point 6. of https://centrifugedao.slack.com/archives/C04HCK98KLL/p1694526988803069?thread_ts=1694518146.332759&cid=C04HCK98KLL
> 
> > When an investment is collected, the invest_amount must be decreased exactly by the CollectedInvestment::::get(&investor, investment_id).amount_payment
> 
> Should this be decreased then?

The `CollectedInvestment::<T>::get(&investor, investment_id).amount_payment` exactly the amount of `pool_currency` which gets collected and converted into tranche tokens. Thus, the `invest_amount` is exactly the post collect `CollectedInvestment::<T>::get(&investor, investment_id).amount_payment + T::Investment::investment(&investor, investment_id)`.

Reply:
> Say I have an unprocessed investment of 1000 USDC and 500 USDC gets collected. Then, handle_collect(500) is executed and thus the invest_amount set to 500, i.e. "decreased" from the previous 1000 to 500. 

Changing a bit the numbers, supposing you collect 600 instead of 500. Then `handle_collect(600)` is executed and the `invest_amount` is set to 600. But if I collect 600 of 1000 I would expect to have 400 as `invest_amount`, this is my concern, but I think it's too late to understand this.

> Since we know that invest_amount must always equal T::Investment::investment(..) after transitioning an InvestState, we can simply overwrite it and simplify the transition instead of decreasing

Ok, I think this is the key what I was not seeing

Reply:
> Changing a bit the numbers, supposing you collect 600 instead of 500. Then `handle_collect(600)` is executed and the `invest_amount` is set to 600

In this case, you are collecting 600 from 1000 and thus the unprocessed investemt is 400 = 1000 - 600 so that you end up calling `handle_collect(400)`.